policy/decompressor.go

```diff
@@ -0,0 +1,133 @@
+package policy
+
+import (
+	"sync"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3metrics/generated/proto/schema"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+)
+
+// Decompressor maintains a mapping of integers to policies. It is used to return a Policy from
+// its corresponding integer encoding
+type Decompressor interface {
+	Get(id int64) (Policy, bool)
+}
+
+type decompressorMap map[int64]Policy
+
+type noopDecompression struct{}
+
+// NewNoopDecompressor returns a new Decompressor mapping which never matches an id
+func NewNoopDecompressor() Decompressor {
+	return noopDecompression{}
+}
+
+func (m noopDecompression) Get(id int64) (Policy, bool) {
+	return EmptyPolicy, false
+}
+
+type staticDecompressor struct {
+	policies decompressorMap
+}
+
+// NewStaticDecompressor returns a new static Decompressor
+func NewStaticDecompressor(policies map[int64]Policy) Decompressor {
+	return staticDecompressor{policies: policies}
+}
+
+func (m staticDecompressor) Get(id int64) (Policy, bool) {
+	p, ok := m.policies[id]
+	return p, ok
+}
+
+type dynamicDecompressor struct {
+	sync.RWMutex
+
+	policies decompressorMap
+	log      xlog.Logger
+}
+
+// NewDynamicDecompressor returns a new dynamic Decompressor
+func NewDynamicDecompressor(w kv.ValueWatch, opts instrument.Options) Decompressor {
+	m := &dynamicDecompressor{
+		policies: make(decompressorMap),
+		log:      opts.Logger(),
+	}
+	m.update(w.Get())
+
+	go m.run(w)
+
+	return m
+}
+
+func (m *dynamicDecompressor) Get(id int64) (Policy, bool) {
+	m.RLock()
+	p, ok := m.policies[id]
+	m.RUnlock()
+	return p, ok
+}
+
+func (m *dynamicDecompressor) run(w kv.ValueWatch) {
+	for {
+		_, ok := <-w.C()
+		if ok {
+			m.log.Debug("received an update to decompressor policies")
+			m.update(w.Get())
+		} else {
+			m.log.Debug("watch for decompressor policies was closed")
+			return
+		}
+	}
+}
+
+func (m *dynamicDecompressor) update(v kv.Value) {
+	if v == nil {
+		m.log.Warn("received a nil Value for decompressor update")
+		return
+	}
+
+	activePolicies := schema.ActivePolicies{}
+	err := v.Unmarshal(&activePolicies)
+	if err != nil {
+		m.log.WithFields(
+			xlog.NewLogField("version", v.Version()),
+			xlog.NewLogErrField(err),
+		).Error("failed to unmarshal decompressor update")
+		return
+	}
+
+	encodedPolicies := activePolicies.GetCompressedPolicies()
+	if encodedPolicies == nil {
+		m.log.Warn("encoded policies in update to decompressor is nil")
+		return
+	}
+
+	newPolicies := make(decompressorMap, len(encodedPolicies))
```

Comment:
Same, perhaps we can move the construction logic into a separate struct (e.g., compressedPolicies), which provides APIs for you look up policies by id and ids by policies.